File: packages/runtime-go/core/oauth2.go
package core

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// OAuth2 token acquisition.
//
// Only the flows an SDK can honestly own (SPEC.md §3.1.6): client credentials, where the SDK holds the
// credentials and is entirely responsible, and refreshing a token the caller obtained elsewhere. The
// authorization-code redirect needs a browser and a human, so it stays the application's job.
//
// The token request is the easy part. What earns this file are the three things around it:
// single-flight refresh, proactive expiry, and retrying a 401 exactly once.
//
// Go's single-flight is shaped differently from the other runtimes. There is no promise to hand a
// second caller, so waiters block on a channel that the one in-flight fetch closes — which also gives
// context cancellation somewhere to hook into, something a plain mutex would not.

// OAuth2Flow identifies which grant the SDK uses.
type OAuth2Flow string

const (
	// FlowClientCredentials is machine-to-machine: the SDK holds the id and secret.
	FlowClientCredentials OAuth2Flow = "client_credentials"
	// FlowRefreshToken keeps a token the caller obtained elsewhere current.
	FlowRefreshToken OAuth2Flow = "refresh_token"
)

// OAuth2ClientAuth is where the client credentials go.
//
// RFC 6749 requires servers to support the header form and says the body form *may* be supported, so
// the header is the default — but real servers get this wrong in both directions, which is why it is
// an option rather than a constant.
type OAuth2ClientAuth string

const (
	ClientAuthBasic OAuth2ClientAuth = "basic"
	ClientAuthBody  OAuth2ClientAuth = "body"
)

// OAuth2Config describes how the SDK obtains a token.
type OAuth2Config struct {
	Flow         OAuth2Flow
	TokenURL     string
	ClientID     string
	ClientSecret string
	RefreshToken string
	Scopes       []string
	// ClientAuth defaults to ClientAuthBasic.
	ClientAuth  OAuth2ClientAuth
	ExtraParams map[string]string
}

// expirySkew is how early a token is treated as expired.
//
// Refreshing only once a token has *already* expired guarantees at least one failed request first.
// Thirty seconds covers ordinary clock skew and the time a request spends in flight.
const expirySkew = 30 * time.Second

// OAuth2Error is returned when the authorization server refuses to issue a token.
//
// Deliberately not an *APIError: this is a different service with different failure semantics, and a
// caller handling a 4xx from the API must not catch it by accident.
type OAuth2Error struct {
	Message string
	// Status from the token endpoint. Zero when it was unreachable.
	Status int
	// Code is the RFC 6749 `error` value, e.g. `invalid_client`.
	Code string
}

func (e *OAuth2Error) Error() string {
	if e.Code != "" {
		return fmt.Sprintf("%s (%s, status %d)", e.Message, e.Code, e.Status)
	}
	return fmt.Sprintf("%s (status %d)", e.Message, e.Status)
}

func (e *OAuth2Error) sdkError() {}

type cachedToken struct {
	accessToken string
	// usableUntil is the zero time when the server declared no expiry.
	usableUntil  time.Time
	refreshToken string
}

func (t *cachedToken) expired() bool {
	return !t.usableUntil.IsZero() && !time.Now().Before(t.usableUntil)
}

// TokenSource acquires and caches OAuth2 tokens.
//
// One instance per client, so a client shared across handlers shares one token rather than fetching
// one per call. Safe for concurrent use.
type TokenSource struct {
	config     OAuth2Config
	httpClient *http.Client

	mu     sync.Mutex
	cached *cachedToken
	// inFlight is non-nil while a fetch is running, and is closed when it finishes.
	//
	// This is the single-flight mechanism. Ten concurrent goroutines on a cold client must produce
	// **one** token request: without it, the first thing a new SDK does under load is hammer the
	// authorization server, and the symptom is unexplained 429s from a host nobody configured.
	//
	// A channel rather than sync.Once, because a fetch can *fail* and must be retriable — Once would
	// permanently cache the failure. A channel rather than holding the mutex across the request,
	// because a waiter needs to be able to abandon the wait when its context is cancelled.
	inFlight chan struct{}
	// fetchErr is the result of the in-flight fetch, readable once inFlight is closed.
	fetchErr error
}

// NewTokenSource builds a TokenSource. A nil httpClient uses http.DefaultClient.
func NewTokenSource(config OAuth2Config, httpClient *http.Client) *TokenSource {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	if config.ClientAuth == "" {
		config.ClientAuth = ClientAuthBasic
	}
	return &TokenSource{config: config, httpClient: httpClient}
}

// Token returns a usable access token, fetching or refreshing only when necessary.
func (s *TokenSource) Token(ctx context.Context) (string, error) {
	s.mu.Lock()
	if s.cached != nil && !s.cached.expired() {
		token := s.cached.accessToken
		s.mu.Unlock()
		return token, nil
	}
	s.mu.Unlock()
	return s.refresh(ctx)
}

// ForceRefresh discards the cached token and fetches a new one.
//
// Called on a 401 as well as on expiry, because expiry arithmetic is necessary but not sufficient —
// clocks disagree and servers revoke tokens early.
func (s *TokenSource) ForceRefresh(ctx context.Context) (string, error) {
	s.mu.Lock()
	s.cached = nil
	s.mu.Unlock()
	return s.refresh(ctx)
}

func (s *TokenSource) refresh(ctx context.Context) (string, error) {
	s.mu.Lock()

	// Re-checked under the lock: a caller that arrived just after another goroutine's fetch completed
	// should use its result rather than starting another.
	if s.cached != nil && !s.cached.expired() {
		token := s.cached.accessToken
		s.mu.Unlock()
		return token, nil
	}

	if wait := s.inFlight; wait != nil {
		s.mu.Unlock()
		select {
		case <-wait:
			// The fetch finished. Read its outcome under the lock.
			s.mu.Lock()
			err := s.fetchErr
			var token string
			if s.cached != nil {
				token = s.cached.accessToken
			}
			s.mu.Unlock()
			if err != nil {
				return "", err
			}
			if token == "" {
				return "", &OAuth2Error{Message: "token refresh produced no token", Status: 0}
			}
			return token, nil
		case <-ctx.Done():
			// Abandoning the wait does not cancel the fetch: another caller may still want it, and a
			// cancelled context here is this caller's decision, not a failure of the token source.
			return "", &ConnectionError{Message: "cancelled while waiting for a token", Cause: ctx.Err()}
		}
	}

	// This goroutine owns the fetch.
	done := make(chan struct{})
	s.inFlight = done
	s.fetchErr = nil
	previous := s.cached
	s.mu.Unlock()

	token, err := s.request(ctx, previous)

	s.mu.Lock()
	if err == nil {
		s.cached = token
	}
	s.fetchErr = err
	s.inFlight = nil
	s.mu.Unlock()
	// Closed after the state is published, so a waiter that wakes sees the result rather than racing
	// with the write.
	close(done)

	if err != nil {
		return "", err
	}
	return token.accessToken, nil
}

func (s *TokenSource) request(ctx context.Context, previous *cachedToken) (*cachedToken, error) {
	form := url.Values{}
	switch s.config.Flow {
	case FlowClientCredentials:
		form.Set("grant_type", "client_credentials")
	case FlowRefreshToken:
		form.Set("grant_type", "refresh_token")
		// The rotated token when the server issued one, otherwise the caller's original. A server
		// that rotates refresh tokens invalidates the old one, so reusing it would fail on the
		// *second* refresh — a bug that only appears after a token lifetime has elapsed.
		refresh := s.config.RefreshToken
		if previous != nil && previous.refreshToken != "" {
			refresh = previous.refreshToken
		}
		if refresh == "" {
			return nil, &OAuth2Error{Message: "no refresh token available", Status: 0}
		}
		form.Set("refresh_token", refresh)
	default:
		return nil, &OAuth2Error{
			Message: fmt.Sprintf("unsupported OAuth2 flow %q", s.config.Flow),
			Status:  0,
		}
	}

	if len(s.config.Scopes) > 0 {
		form.Set("scope", strings.Join(s.config.Scopes, " "))
	}
	for key, value := range s.config.ExtraParams {
		form.Set(key, value)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.config.TokenURL,
		strings.NewReader(form.Encode()))
	if err != nil {
		return nil, &OAuth2Error{Message: "building the token request: " + err.Error(), Status: 0}
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	if s.config.ClientID != "" {
		if s.config.ClientAuth == ClientAuthBasic {
			credentials := s.config.ClientID + ":" + s.config.ClientSecret
			req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(credentials)))
		} else {
			form.Set("client_id", s.config.ClientID)
			if s.config.ClientSecret != "" {
				form.Set("client_secret", s.config.ClientSecret)
			}
			// Rebuilt because the body was already encoded above.
			req.Body = io.NopCloser(strings.NewReader(form.Encode()))
			req.ContentLength = int64(len(form.Encode()))
		}
	}

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, &OAuth2Error{
			Message: "could not reach the token endpoint at " + s.config.TokenURL,
			Status:  0,
		}
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(io.LimitReader(resp.Body, maxErrBody))

	var payload struct {
		AccessToken      string  `json:"access_token"`
		RefreshToken     string  `json:"refresh_token"`
		ExpiresIn        float64 `json:"expires_in"`
		Error            string  `json:"error"`
		ErrorDescription string  `json:"error_description"`
	}
	_ = json.Unmarshal(body, &payload)

	if resp.StatusCode >= 400 {
		// Never retried. A 400 from a token endpoint means the credentials are wrong, and retrying it
		// is both pointless and indistinguishable from a brute-force attempt.
		message := payload.ErrorDescription
		if message == "" {
			message = payload.Error
		}
		if message == "" {
			message = fmt.Sprintf("the token endpoint returned %d", resp.StatusCode)
		}
		return nil, &OAuth2Error{Message: message, Status: resp.StatusCode, Code: payload.Error}
	}

	if payload.AccessToken == "" {
		return nil, &OAuth2Error{
			Message: "the token endpoint returned no access_token",
			Status:  resp.StatusCode,
		}
	}

	token := &cachedToken{accessToken: payload.AccessToken, refreshToken: payload.RefreshToken}
	if token.refreshToken == "" && previous != nil {
		// A server that does not rotate keeps the previous value usable.
		token.refreshToken = previous.refreshToken
	}
	if payload.ExpiresIn > 0 {
		lifetime := time.Duration(payload.ExpiresIn * float64(time.Second))
		if lifetime > expirySkew {
			lifetime -= expirySkew
		} else {
			lifetime = 0
		}
		token.usableUntil = time.Now().Add(lifetime)
	}
	return token, nil
}

File: packages/runtime-dotnet/src/OAuth2.cs
using System.Text;

namespace Graft.Runtime;

/// <summary>Configuration for an OAuth2 token source (SPEC.md §3.1.6).</summary>
public sealed record OAuth2Config
{
  public required string TokenUrl { get; init; }

  public string? ClientId { get; init; }

  public string? ClientSecret { get; init; }

  public string? RefreshToken { get; init; }

  public IReadOnlyList<string> Scopes { get; init; } = Array.Empty<string>();
}

/// <summary>
/// Fetches and refreshes an access token.
/// </summary>
/// <remarks>
/// <para>
/// <b>Single-flight, under a <see cref="SemaphoreSlim"/>.</b> A client is shared across requests and .NET has
/// real concurrency, so two callers discovering an expired token at once would both refresh — spending two token
/// requests and, with providers that invalidate the previous token, breaking one of them. The cache is
/// re-checked <i>inside</i> the lock, because a caller that waited must not refresh on the basis of what it saw
/// before waiting.
/// </para>
/// <para>
/// <see cref="SemaphoreSlim"/> rather than <c>lock</c>, because the critical section awaits: <c>lock</c> cannot
/// span an <c>await</c> in C#, and the language enforces that rather than leaving it as a subtle bug.
/// </para>
/// </remarks>
public sealed class TokenSource
{
  /// <summary>Refresh this far before expiry, because clocks disagree and a token in flight can expire.</summary>
  private static readonly TimeSpan ExpirySkew = TimeSpan.FromSeconds(30);

  /// <summary>What a provider that omits <c>expires_in</c> gets. Treating it as "never" would cache a dead token.</summary>
  private static readonly TimeSpan DefaultLifetime = TimeSpan.FromHours(1);

  private readonly OAuth2Config _config;
  private readonly ITransport _transport;
  private readonly TimeSpan _timeout;
  private readonly Func<DateTimeOffset> _clock;
  private readonly SemaphoreSlim _gate = new(1, 1);

  private string? _accessToken;
  private DateTimeOffset? _expiresAt;

  public TokenSource(
      OAuth2Config config,
      ITransport transport,
      TimeSpan? timeout = null,
      Func<DateTimeOffset>? clock = null)
  {
    _config = config;
    _transport = transport;
    _timeout = timeout ?? TimeSpan.FromSeconds(30);
    _clock = clock ?? (() => DateTimeOffset.UtcNow);
  }

  public async Task<string> TokenAsync(CancellationToken cancellationToken = default)
  {
    await _gate.WaitAsync(cancellationToken).ConfigureAwait(false);
    try
    {
      var now = _clock();
      if (_accessToken is not null && _expiresAt is not null && now < _expiresAt.Value - ExpirySkew)
      {
        return _accessToken;
      }

      return await FetchAsync(cancellationToken).ConfigureAwait(false);
    }
    finally
    {
      _gate.Release();
    }
  }

  /// <summary>Drop the cached token, so the next call fetches a fresh one. Used by the 401-retry path.</summary>
  public void Invalidate()
  {
    _accessToken = null;
    _expiresAt = null;
  }

  private async Task<string> FetchAsync(CancellationToken cancellationToken)
  {
    var form = new Dictionary<string, string>();
    if (_config.RefreshToken is not null)
    {
      form["grant_type"] = "refresh_token";
      form["refresh_token"] = _config.RefreshToken;
    }
    else
    {
      form["grant_type"] = "client_credentials";
    }

    if (_config.Scopes.Count > 0)
    {
      form["scope"] = string.Join(" ", _config.Scopes);
    }

    var headers = new Dictionary<string, string>
    {
      ["Content-Type"] = "application/x-www-form-urlencoded",
      ["Accept"] = "application/json",
    };

    // Credentials in the Authorization header when both are present: that is the form every provider
    // accepts, where in-body credentials are optional in the spec and unevenly implemented.
    if (_config.ClientId is not null && _config.ClientSecret is not null)
    {
      var credentials = Convert.ToBase64String(
          Encoding.UTF8.GetBytes($"{_config.ClientId}:{_config.ClientSecret}"));
      headers["Authorization"] = "Basic " + credentials;
    }
    else if (_config.ClientId is not null)
    {
      form["client_id"] = _config.ClientId;
    }

    var encoded = string.Join(
        "&",
        form.Select(pair => $"{Uri.EscapeDataString(pair.Key)}={Uri.EscapeDataString(pair.Value)}"));

    var response = await _transport
        .SendAsync(
            new HttpRequestSpec(
                "POST",
                _config.TokenUrl,
                headers,
                encoded,
                new Dictionary<string, IReadOnlyList<string>>()),
            _timeout,
            cancellationToken)
        .ConfigureAwait(false);

    if (response.StatusCode is < 200 or >= 300)
    {
      // Never retried: a 400 from a token endpoint means the credentials are wrong, and retrying wrong
      // credentials is how an account gets locked.
      var detail = response.Body.Length > 500 ? response.Body[..500] : response.Body;
      throw new OAuth2Exception($"token request failed with {response.StatusCode}: {detail}");
    }

    if (Json.Parse(response.Body) is not IDictionary<string, object?> map
        || !map.TryGetValue("access_token", out var raw)
        || raw is not string token)
    {
      throw new OAuth2Exception("token response had no string access_token");
    }

    _accessToken = token;
    map.TryGetValue("expires_in", out var expiresIn);
    var lifetime = expiresIn switch
    {
      long seconds => TimeSpan.FromSeconds(seconds),
      double seconds => TimeSpan.FromSeconds(seconds),
      _ => DefaultLifetime,
    };
    _expiresAt = _clock() + lifetime;
    return token;
  }
}

/// <summary>OAuth2, holding a token source that refreshes itself.</summary>
public sealed class OAuth2Auth : IAuth
{
  private readonly TokenSource _source;

  public OAuth2Auth(TokenSource source) => _source = source;

  public async Task<AuthApplied> ApplyAsync(
      IReadOnlyDictionary<string, string> headers,
      IReadOnlyDictionary<string, IReadOnlyList<string>> query,
      CancellationToken cancellationToken = default)
  {
    var token = await _source.TokenAsync(cancellationToken).ConfigureAwait(false);
    var result = new Dictionary<string, string>(headers) { ["Authorization"] = "Bearer " + token };
    return new AuthApplied(result, query);
  }

  /// <summary>Drop the cached token. The 401-retry path calls this before its one retry.</summary>
  public void Invalidate() => _source.Invalidate();
}
